core/src/wasm.rs

```diff
@@ -0,0 +1,384 @@
+
+use crate::{
+    smartcontracts::{Execute, ValidQuery},
+    wsv::{WorldStateView, WorldTrait},
+};
+
+const WASM_ALLOC_FN: &str = "alloc";
+const WASM_MEMORY_NAME: &str = "memory";
+const WASM_MAIN_FN_NAME: &str = "main";
+const EXECUTE_ISI_FN_NAME: &str = "execute_isi";
+const EXECUTE_QUERY_FN_NAME: &str = "execute_query";
+
+/// `WebAssembly` execution error type
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Runtime initialization failure")]
+    Initialization(#[source] anyhow::Error),
+    #[error("Module instantiation failure")]
+    Instantiation(#[source] anyhow::Error),
+    #[error("Named export not found")]
+    ExportNotFound(#[source] anyhow::Error),
+    #[error("Fn call failed")]
+    FnCall(#[source] Trap),
+    /// Some other error happened
+    #[error("Some other error happened")]
+    Other(#[source] eyre::Error),
+}
+
+struct State<'a, W: WorldTrait> {
+    wsv: &'a WorldStateView<W>,
+    account_id: AccountId,
+
+    /// Number of instructions in the smartcontract
+    instruction_count: u64,
+}
+
+impl<'a, W: WorldTrait> State<'a, W> {
+    fn new(wsv: &'a WorldStateView<W>, account_id: AccountId) -> Self {
+        Self {
+            wsv,
+            account_id,
+            instruction_count: 0,
+        }
+    }
+}
+
+/// `WebAssembly` virtual machine
+pub struct Runtime<'a, W: WorldTrait> {
+    engine: Engine,
+    linker: Linker<State<'a, W>>,
+}
+
+impl<'a, W: WorldTrait> Runtime<'a, W> {
+    /// Every WASM instruction costs approximately 1 unit of fuel. See
+    /// [`wasmtime` reference](https://docs.rs/wasmtime/0.29.0/wasmtime/struct.Store.html#method.add_fuel)
+    const FUEL_LIMIT: u64 = 10_000;
+
+    fn create_config() -> Config {
+        let mut config = Config::new();
+        config.consume_fuel(true);
+        //config.cache_config_load_default();
+        config
+    }
+
+    fn create_engine(config: &Config) -> Result<Engine, Error> {
+        Engine::new(config).map_err(Error::Initialization)
+    }
+
+    fn create_linker(engine: &Engine) -> Result<Linker<State<'a, W>>, Error> {
+        let mut linker = Linker::new(engine);
+
+        linker
+            .func_wrap(
+                "iroha",
+                EXECUTE_ISI_FN_NAME,
+                |mut caller: Caller<State<_>>, offset: u32, len: u32| {
+                    let memory = Self::get_memory(&mut caller)?;
+
+                    // Accessing memory as a byte slice to avoid the use of unsafe
+                    let isi_mem_range = offset as usize..(offset + len) as usize;
+                    let mut isi_bytes = &memory.data(&caller)[isi_mem_range];
+                    let instruction = Instruction::decode(&mut isi_bytes)
+                        .map_err(|error| Trap::new(error.to_string()))?;
+
+                    instruction
+                        .execute(caller.data().account_id.clone(), caller.data().wsv)
+                        .map_err(|error| Trap::new(error.to_string()))?;
```

Comment:
As I understand permissions are not checked here. We should also add these in the future PRs, better to create an issue for it.

Reply:
Opened issue